parsers/PA.py

```diff
@@ -211,69 +175,37 @@ def fetch_production(
             continue
         production_mode = PRODUCTION_TYPE_TO_PRODUCTION_MODE[production_type]  # hydro
         production_mix.add_value(production_mode, float(production_value))
 
-    # Calculate manually thermal generation to avoid using outdated chart data
-    thermal_generation = sum_thermal_units(soup)
-    # Known fossil plants: parse, subtract from "unknown", add to "coal"/"oil"/"gas"
-    thermal_production_breakdown = soup.find_all("table", {"class": "sitr-table-gen"})[
-        1
-    ]
-    # Make sure the table header is indeed "Térmicas (MW)" (in case the tables are re-arranged)
-    thermal_production_breakdown_table_header = (
-        thermal_production_breakdown.parent.parent.parent.select("> .tile-title")[
-            0
-        ].string
-    )
-    assert "Térmicas" in thermal_production_breakdown_table_header, (
-        f"Exception when extracting thermal generation breakdown for {zone_key}: table header does not contain "
-        f"'Térmicas' but is instead named {thermal_production_breakdown_table_header}"
-    )
-    thermal_production_units = thermal_production_breakdown.select(
-        "tbody tr td table.sitr-gen-group tr"
-    )
-
-    for thermal_production_unit in thermal_production_units:
-        unit_name_and_generation = thermal_production_unit.find_all("td")
-        unit_name = unit_name_and_generation[0].string
-        unit_generation = float(unit_name_and_generation[1].string)
-        if unit_name in MAP_THERMAL_GENERATION_UNIT_NAME_TO_FUEL_TYPE:
-            if unit_generation > 0:  # Ignore self-consumption
-                unit_fuel_type = MAP_THERMAL_GENERATION_UNIT_NAME_TO_FUEL_TYPE[
-                    unit_name
-                ]
-                production_mix.add_value(unit_fuel_type, unit_generation)
-                thermal_generation -= unit_generation
-        else:
-            logger.warning(
-                f"{unit_name} is not mapped to generation type",
-                extra={"key": zone_key},
-            )
-
-    # Round remaining "unknown" output to 13 decimal places to get rid of floating point errors
-    thermal_generation = round(thermal_generation, 13)
-    if 0 < thermal_generation < 1e-3:
-        thermal_generation = 0.0
-    # assign anything remaining to 'unknown'
-    production_mix.add_value("unknown", max(thermal_generation, 0.0))
+    # calculate our own 'termica' data by summing known fossil plants and bucketing into "coal"/"oil"/"gas"
+    production_units = response_payload["unit"]
+    for production_unit in production_units:
+        unit_name = production_unit["name"]
+        unit_generation = float(production_unit["value"])
+        if (
+            unit_name in MAP_THERMAL_GENERATION_UNIT_NAME_TO_FUEL_TYPE
+            and unit_generation >= 0
+        ):  # Ignore self-consumption
+            unit_fuel_type = MAP_THERMAL_GENERATION_UNIT_NAME_TO_FUEL_TYPE[unit_name]
+            production_mix.add_value(unit_fuel_type, unit_generation)
```

Comment:
note: maybe the only downside is that now we cannot know dynamically which plants are "thermic" plants (because there is no way to know from the json which units are displayed on the webpage under the "Termicas" table). So if a new unknown fossil plant was to be built we wouldn't sum it up / get a warning. So far though we "know" of all existing plants 🙌🏽

Reply:
Unfortunately the data is coming down like this https://sitr.cnd.com.pa/m/pub/data/gen.json , where `"units"` is just a long list of energy plant names and associated production - without any information about which plants are "Thermica", which are solar, which are hydroelectric, etc...

The grouping into production modes is introduced only at the frontend level https://sitr.cnd.com.pa/m/pub/gen.html (couldn't quite figure out if they had a dynamic way to render that or if essentially they are hard coding which plant goes into which table) so I fear that there is not really a way to be able to know which plants are thermic without either "knowing it" or scraping the html page...

Happy to revert if risking to miss out on new plants is a blocker!